### DISCUSSION TITLE
v4.1.6

### DISCUSSION THREAD

**Discussion by autoSteve** [BJReplay/ha-solcast-solar#164] [Announcements]
## What's Changed
A change occurred in HA 2024.9.0 that caused an issue in the `configure` initial dialogue for this integration. So it's fixed, but it's even better. Now most everything is on a single dialogue, with only manual dampening factors a separate configuration.

* Simplify configure dialogue by @autoSteve

**Prior beta changes in v4.1.4 and v4.1.5**:
* Update Polish translation by @home409ca
* Rename integration in HACS to Solcast PV Forecast by @BJReplay
* Reduce aiofiles version requirement to >=23.2.0 by @autoSteve
* Configuration dialog improvements by @autoSteve
* Misc translation updates by @autoSteve
* Refactor moment and remaining spline build by @autoSteve
* Prevent negative forecast for X hour sensor by @autoSteve
* Suppress spline bounce for reducing spline by @autoSteve
* More careful serialisation of solcast.json by @autoSteve
* Monitor last updated timestamp for sites-usage.json by @autoSteve
* Extensive code clean-up by @autoSteve
* Bug: Timestamp stored in usage cache was wrong by @autoSteve
* Bug: Adding API key reset usage for first key by @autoSteve
* Bug: Missing iterator in new sites check by @autoSteve
* Work around a possible HA scheduling bug by @autoSteve
* Code style alignment to HA style guidelines by @autoSteve

Full Changelog: https://github.com/BJReplay/ha-solcast-solar/compare/v4.1.3...v4.1.6

**Comment by autoSteve** [comment]
> state_class comment made on 4.1.4 (but seen first on 4.1.3) https://github.com/BJReplay/ha-solcast-solar/discussions/150#discussioncomment-10513137

Tracking here, @gcoan.

**Reply by gcoan** [reply]
😁

**Reply by autoSteve** [reply]
> state_class comment made on 4.1.4 (but seen first on 4.1.3) [#150 (reply in thread)](https://github.com/BJReplay/ha-solcast-solar/discussions/150#discussioncomment-10513137)

Sooo... Geoff, @gcoan, luv.

Do you think this really needs poking? Most of the sensors do not even specify a state class.

To quote Aussie "strine" in full voice, "Yeah, nah cobber, I reckon it doesn't Parramatta to no bugger, 'cause it don't break sh\*t. Me Home Arse-siss ain't busted... How's your sh\*t?" (Translation: I think no one cares because it does not break anything in Home Assistant. What is your installation doing? Is it okay?)

Yeah, it makes your eye twitch, but it's all good from what I can see.

Is it all good for you? Apart from the inconsistency? i.e., "How's your sh*t?"

**Reply by gcoan** [reply]
My eye is indeed twitching, not least of which due to the inconsistency across what should be identical sensors, but also what the longer term impacts of having these things wrong or missing.

I read https://developers.home-assistant.io/docs/core/entity/sensor/#long-term-statistics several times and this does suggest how Home Assistant does various statistics gathering calculations based upon what the state class is set to. So presumably this means some sensors are being aggregated over time incorrectly, and others not at all that should be.

But it also appears that there are bears out there and the importance of choosing the state class correctly is highlighted.  This suggests (didn't google for evidence) that changing the class may have consequences you didn't anticipate.

So potentially bad if we don't change and different bad if we do.

And as you say at this point no-one has complained about it.

I think it would be worth doing a test on a test rig to see what the impact of this is if we do correct the state classes as I suggested needed to be, and then we have some information that we don't currently have.  At that point I could acquiesce and this be parked in the 'known not quite right' category with some evidence that its 'too hard/bad' or 'not bad but no obvious major justification to change'.

Information is king and at the moment we are veering to not opening the bear cage because we fear what's inside or we can't be bothered finding out.  Once we count the bears 🐻 we may still not go inside but we at least know what level of armour we would need should we ever want to go bear-hunting.

Analogies posted, ball hit back to the other side of the court 🎾 waiting for the return volley

**Reply by autoSteve** [reply]
"Jeez, Guv. Whoah! Propper 'send' you-know-English type lob that was! Sheesh! Almost didn't get 'im!" (returns volley unenthusiastically with a limp wrist but it hits its target...)

Okay, I'll set up a test rig...

Ugh. This is gonna hurt.

But you're right. And the pedant in me knows it.

**Reply by autoSteve** [reply]
I did not do this, and yet I am somehow inexplicably on the hook to un-f\*ck it up. The story of my life.

A low priority, if you'll allow @gcoan.

**Reply by gcoan** [reply]
> I did not do this, and yet I am somehow inexplicably on the hook to un-f*ck it up. The story of my life.

I remain amazed at how you have taken on this integration as if it were your own baby.  You have become a super-star foster parent

> A low priority, if you'll allow @gcoan.

Of course

Said the horse

🐴

**Reply by autoSteve** [reply]
For those wishing to follow along, the branch Post-v4.1.6 at BJReplay explicitly sets a state_class for every sensor.

I am running this code in prod, and it is committed. So I am commited. I have just sent it for the good of community. I could not be arsed setting up a play pen to see what happened, because I suspect the answer will be, "Meh. Not much...", and a real live dev set-up would have impacted my prod severely given limitations on API calls.

(If you're listening Solcast peeps, @billy-solcast then here's one for your suggestions box... How about a programme for genuine registered developers that allows more than a measly ten calls per day without violating your terms of agreement by using multiple email addresses to get multiple API keys??? I don't do that, by the way, so I suffer, and I am slowed down in development. But developers get it. We do our best to not flood you with requests because efficiency. We try to be the reason that many thousands of people do not flood you. But sometimes we _need_ a little space for creativity, like running three servers at once working three different problems... Even the fifty calls of old would be gold.)

@gcoan, the worst that will happen is that I will lick my wounds and wind back the branch commit and restore from backup.

But for now I am just sending it because API limits. I'll pick up the pieces later. Reviewing the fallout tomorrow.

**Reply by gcoan** [reply]
> If you're listening Solcast peeps, @billy-solcast then here's one for your suggestions box... How about a programme for genuine registered developers that allows more than a measly ten calls per day without violating your terms of agreement

I am pretty sure that there are some pre-defined locations in the API specification that you can get the forecast for that are free to call.

Ha, found it https://docs.solcast.com.au/#unmetered-locations

**Reply by autoSteve** [reply]
Fair @gcoan, but that is targeted at testers, not developers and supporters of developed code, who probably have a corporate contract and can do what they like.

It's just a lousy few more API calls for "power users" / "developers" (that I'd be willing to pay for) is all that would be required to spin up more dev instances. You get how much I love this baby @gcoan. But I can't support it well at times. If I spin up two more instances then you'd think I am at only three-per-day max per instance for the three total. But configs need to be changed in all, including my **prod** instance. Prod gets the bulk (six calls), everything else gets just _two_ call per day, which are usually exhausted in short order. Dev is slowed down, or like this evening I just decide, "I've had enough. Just send it to production." This does not sit well with me, working at a bank, and the regulator APRA would have kittens, but it is what it is. And this is a missed revenue opportunity, Solcast...

This results in slower development for our community. I have prod usually at eight calls. Then I have **one** dev, which is very much a second-class citizen that gets a couple of API calls per day.

I guess that everybody I support now must live at the Taj Mahal... Pack ya virtual bags, because you're being spun up in Agra folks (with much hacking to do so), but not so much real world.

**Reply by autoSteve** [reply]
The state_class changes have been rolled back for now.

This change would opt-in those sensors to LTS, and I am not convinced that this is a good thing, so will test it over the longer term.

**Reply by gcoan** [reply]
The LTS side effect doesn't surprise me.  No issues with rolling back, we need to understand how this change affects everything

**Comment by gcoan** [comment]
Curiously HACS hasn't prompted me to upgrade despite there being a new version. Didn't get prompted for the last release but thought I had fixed that.
Will give it a bit and see if it works it out
![image](https://github.com/user-attachments/assets/d0bacca9-1dd3-491a-9dd2-86abf0fabaf2)

Anyway, feedback item to raise that the readme page will need updating for the new configuration dialogue
![image](https://github.com/user-attachments/assets/e0017d86-5bce-4720-bf95-526987534797)

**Reply by autoSteve** [reply]
With HACS 1.x updates were checked every three days, when one restarted HA, or when updated information was manually requested (I think those were the only circumstances).

I'm not sure about the behaviour of HACS 2.x.

Some text has subtly changed, with that initial configuration dialogue now reading, "API limit (optionally comma separate multiple values for each key)". That all?

**Reply by gcoan** [reply]
I thought HACS 1.x polled every day but only once a day 🤷‍♂️

Was poking around in HACS and all of a sudden it noticed that there was an update for solcast and predbat.  Can't remember if this was before or after I restarted HA so no conclusions there.

Haven't yet upgraded to the new version but I thought that the configuration was now all in one large dialogue so the sequence of screens in the readme were out of date now?

**Reply by autoSteve** [reply]
The initial config (when it is first installed) is just the key and limit, and that's unchanged from how it was.

Subsequent `configure` dialogue used to be a list (that got broke), with sub-pages, and that is what's been replaced.

You can see the new dialogue further down in the readme where dampening and other options are chatted about.

**Reply by gcoan** [reply]
Jolly good. Thanks for the clarification
